modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Traverse.kt

```diff
@@ -5,11 +5,519 @@ import arrow.core.Id
+ * import arrow.fx.extensions.io.concurrent.parTraverse
+ * import arrow.fx.extensions.io.unsafeRun.runBlocking
+ * import arrow.unsafe
+ *
+ * interface Profile
+ * interface User
+ * //sampleStart
+ * data class DummyUser(val name: String) : User
+ * data class DummyProfile(val u: User) : Profile
+ *
+ * fun userInfo(u: User): IO<Profile> =
+ *   IO { DummyProfile(u) } // this can be a call to the DB
+ *
+ * fun List<User>.processLogin(): IO<List<Profile>> =
+ *   parTraverse { userInfo(it) }
+ *
+ * fun program(): IO<Unit> = IO.fx {
+ *   val list = listOf(DummyUser("Micheal"), DummyUser("Juan"), DummyUser("T'Challa"))
+ *     .processLogin().bind()
+ *   println(list)
+ * }
+ * //sampleEnd
+ * fun main() {
+ *   unsafe { runBlocking { program() } }
+ * }
+ * ```
+ *
+ * Evidently, [Traverse] is not limited to [List] or [Nel], it provides an abstraction over 'things that can be traversed over', like a Binary tree, a [SequenceK], or a `Stream`, hence the name [Traverse].
+ *
+ * #### Playing with `Reader`
+ *
+ * Another interesting data type we can use is `Reader`. Recall that a `Reader<D, A>` is a type alias for `Kleisli<ForId, D, A>` which is a wrapper around `(D) -> A`.
+ *
+ * If we fix `D` to be some sort of dependency or configuration and `A` as the return type, we can use the `Reader` applicative in our [traverse].
+ *
+ * ```kotlin:ank
+ * import arrow.mtl.Reader
+ *
+ * interface Context
+ * interface Topic
+ * interface Result
+ *
+ * typealias Job<A> = Reader<Context, A>
+ *
+ * fun processTopic(t: Topic): Job<Result> = TODO()
+ * ```
+ *
+ * We can imagine we have a data pipeline that processes a bunch of data, each piece of data being categorized by a topic. Given a specific topic, we produce a `Job` that processes that topic. (Note that since a `Job` is just a `Reader`/`Kleisli`, one could write many small `Jobs` and compose them together into one `Job` that is used/returned by `processTopic`.)
+ *
+ * Corresponding to bunch of topics, a `List<Topic>` if you will. Since `Reader` has an [Applicative] instance, we can [traverse] over this list with `processTopic`.
+ *
+ * ```kotlin:ank
+ * import arrow.core.Id
+ * import arrow.core.ForId
+ * import arrow.core.ListK
+ * import arrow.core.extensions.id.applicative.applicative
+ * import arrow.mtl.KleisliPartialOf
+ * import arrow.mtl.extensions.kleisli.applicative.applicative
+ * import arrow.mtl.fix
+ *
+ * val JobForContext = Job.applicative<ForId, Context>(Id.applicative())
+ *
+ * fun processTopics(topics: ListK<Topic>): Job<ListK<Result>> =
+ *   topics.traverse<KleisliPartialOf<ForId, Context>, Result>(JobForContext) {
+ *     processTopic(it)
+ *   }.fix()
+ * ```
+ *
+ * Note the nice return type - `Job<List<Result>>`. We now have one aggregate `Job` that when run, will go through each topic and run the topic-specific job, collecting results as it goes. We say "when run" because a `Job` is some function that requires a Context before producing the value we want.
```

Comment:
```suggestion
 * Note the nice return type - `Job<List<Result>>`. We now have one aggregate `Job` that, when run, will go through each topic and run the topic-specific job, collecting results as it goes. We say "when run" because a `Job` is some function that requires a Context before producing the value we want.
```